File: cpp/76.cpp
#include <iostream>
#include <vector>
#include <algorithm>
#include <iterator>
#include <utility>
#include <map>
#include <set>
#include <string>
#include <unordered_map>
#include <climits>
#include <numeric>
using namespace std;

template<typename T>
void prints(vector<T> &v){
  cout << "[";
  for(uint i=0; i<v.size()-1; ++i)
    cout << v[i] << ", ";
  if(v.size()) cout << v[v.size()-1];
  cout << "]\n";
}

class Solution {
public:
  string minWindow(string S, string T) {
    map<char, int> tcnt;
    for(auto &c : T) ++tcnt[c];
    map<char, int> scnt;
    int l = 0, r = 0, todo = tcnt.size(),
      ans = INT_MAX, ansl = -1, ansr = -1;
    while(r < S.length()){
      ++scnt[S[r]];
      if(scnt[S[r]] == tcnt[S[r]])
        --todo;
      ++r;
      if(todo == 0){
        while(scnt[S[l]] > tcnt[S[l]])
          --scnt[S[l++]];
        if(r-l < ans)
          ans = r-l, ansl = l, ansr = r;
        --scnt[S[l++]]; ++todo;
      }
    }
    if(ans != INT_MAX)
      return S.substr(ansl, ansr - ansl);
    else
      return "";
  }
};

int main(){
  auto sol = Solution();
  cout << sol.minWindow("a", "aa");
}

File: python/76.py
class Solution(object):
    def minWindow(self, s, t):
        """
        :type s: str
        :type t: str
        :rtype: str
        """
        pat = collections.Counter(t)
        l = r = 0
        cnt = 0
        cur = collections.defaultdict(int)
        ans = None

        while l < len(s):
            while r < len(s) and cnt < len(pat):
                cur[s[r]] += 1
                if s[r] in pat and cur[s[r]] == pat[s[r]]:
                    cnt += 1
                r += 1

            # print (l, r, s[l:r], cur)
            if cnt == len(pat):
                while s[l] not in pat or cur[s[l]] > pat[s[l]]:
                    cur[s[l]] -= 1
                    l += 1
                if not ans or len(ans) > r-l:
                    ans = s[l:r]

            cur[s[l]] -= 1
            cnt -= 1
            l += 1

        return '' if ans is None else ans
